### PR TITLE
Add ticked spinlock

### PR DESCRIPTION
**Pull request by ninolomata**
This commit replaces the current spinlock implementation to a ticked spinlock.

### CODE REVIEW

**Comment by josecm** [comment]
I've taken the liberty of adding an implementation for the ticket lock for aarch64. So that we follow the same logic for all the supported architectures, @AfonsoSantos96 could you implement this for aarch32?

Still need to test this more thoroughly. @ninolomata did you test the risc-v implementation this in CVA6?

**Comment by ninolomata** [comment]
No, but I can do it.  I  will also try on the Openpiton because we were having some troubles there with the spinlocks.

**Comment by AfonsoSantos96** [comment]
> I've taken the liberty of adding an implementation for the ticket lock for aarch64, so that we follow the same logic for all the supported architectures. @AfonsoSantos96 could you implement this for aarch32?

Sure, I will carry out this implementation to aarch32.

**Comment by josecm** [comment]
> No, but I can do it. I will also try on the Openpiton because we were having some troubles there with the spinlocks.

That'd be great!

> Sure, I will carry out this implementation to aarch32.

Thanks!

**Comment by ninolomata** [comment]
Just tested on both it's working.

**Comment by josecm** [comment]
> Just tested on both it's working.

Thanks! Leave it with me and @AfonsoSantos96  to wrap up the PR.

**Comment by josecm** [comment]
@AfonsoSantos96 thanks for the commit. Where did you test this? Just the fvp model? Or any real hardware?

File: src/arch/armv8/aarch32/inc/arch/spinlock.h [hunk 1]
```diff
@@ -8,30 +8,62 @@
...
+        "cmp  %r2, #0\n\t"
         "bne 1b \n\t"
-        : "=&r"(tmp), "+Q"(*lock)
-        : "r"(ONE) : "memory");
+        /* Wait for your turn */
+        "2:\n\t"
+        "ldaex  %r1, %4\n\t"
+        "cmp    %r0, %r1\n\t"
+        "beq   3f\n\t"
+        "b 2b\n\t"
```
Comment (DavidMCerdeira) [inline src/arch/armv8/aarch32/inc/arch/spinlock.h:50]: is there a reason for not using wfe here?

File: src/arch/armv8/aarch64/inc/arch/spinlock.h [hunk 2]
```diff
@@ -3,38 +3,67 @@
...
 static inline void spin_unlock(spinlock_t* lock)
 {
-    asm volatile("stlr wzr, %0\n\t" ::"Q"(*lock) : "memory");
+    uint32_t temp;
+
+    __asm__ volatile(
+        /* increment to next ticket */
+        "ldr    %w0, %1\n\t"
+        "add    %w0, %w0, 1\n\t"
+        "stlr   %w0, %1\n\t"
```
Comment (DavidMCerdeira) [inline src/arch/armv8/aarch64/inc/arch/spinlock.h:63]: There should be a `sev` here no?

File: src/arch/riscv/inc/arch/spinlock.h [hunk 3]
```diff
@@ -8,29 +8,48 @@
...
 
 static inline void spin_lock(spinlock_t* lock)
 {
-    spinlock_t const ONE = 1;
-    spinlock_t tmp = SPINLOCK_INITVAL;
+    uint32_t const INCR = 1;
+    uint32_t ticket;
+    uint32_t serving;
 
     asm volatile("1:\n\t"
```
Comment (DavidMCerdeira) [inline src/arch/riscv/inc/arch/spinlock.h:30]: Why have `1:` here?

**Comment by DavidMCerdeira** [comment]
Although it might be obvious in this case, it would be nice to elaborate a bit on why this change is necessary/useful

File: src/arch/armv8/aarch32/inc/arch/spinlock.h (hunk 1 above)
Comment (josecm) [reply src/arch/armv8/aarch32/inc/arch/spinlock.h:50]: Added.

File: src/arch/armv8/aarch64/inc/arch/spinlock.h (hunk 2 above)
Comment (josecm) [reply src/arch/armv8/aarch64/inc/arch/spinlock.h:63]: Added

**Review by josecm — DISMISSED** [review]

**Review by AfonsoSantos96 — APPROVED** [review]

**Review by josecm — APPROVED** [review]

**Pull request merged**